src/tests/test_utils.cpp

```diff
@@ -47,6 +51,81 @@ class Utility_Function_Tests final : public Test {
       }
 
    private:
+      Test::Result test_checked_add() {
+         Test::Result result("checked_add");
+
+         const size_t large = static_cast<size_t>(-5);
+         const size_t zero = 0;
+
+         for(size_t i = 0; i != 5; ++i) {
+            auto sum1 = Botan::checked_add<size_t>(i, zero, zero, zero, large);
+            auto sum2 = Botan::checked_add<size_t>(large, zero, zero, zero, i);
+
+            result.confirm("checked_add looks at all args", sum1 == sum2);
+
+            if(i < 5) {
+               result.test_eq("checked_add worked", sum1.value(), i + large);
+            } else {
+               result.confirm("checked_add did not return a result", !sum1.has_value());
+            }
+         }
```

Comment:
That won't ever hit the failure case, no?

```suggestion
         for(size_t i = 0; i <= 5; ++i) {
            auto sum1 = Botan::checked_add<size_t>(i, zero, zero, zero, large);
            auto sum2 = Botan::checked_add<size_t>(large, zero, zero, zero, i);

            result.confirm("checked_add looks at all args", sum1 == sum2);

            if(i < 5) {
               result.test_eq("checked_add worked", sum1.value(), i + large);
            } else {
               result.confirm("checked_add did not return a result", !sum1.has_value());
            }
         }
```

Reply:
Fixed! This turned out to expose the fact that `checked_add` did not behave correctly. The check was

`if(r - b != a)`

but if `a + b` overflowed in one direction, the subtraction would just overflow back...

I wrote it in mental analogy to * vs /, where that works.